src/iter/mod.rs

```diff
@@ -2169,6 +2176,39 @@ impl<T: ParallelIterator> IntoParallelIterator for T {
 ///
 /// **Note:** Not implemented for `u64`, `i64`, `u128`, or `i128` ranges
 pub trait IndexedParallelIterator: ParallelIterator {
+    /// Normally, parallel iterators are recursively divided into tasks in parallel.
+    /// This adaptor changes the default behavior by splitting the iterator into a **sequence**
+    /// of parallel iterators of given `sizes`.
+    ///
+    /// This can have many applications but the most notable ones are:
+    /// - better performances with [`find_first()`]
+    /// - more predictable performances with [`find_any()`]
+    /// - better locality (especially if the reduce operation re-use folded data)
+    ///
+    /// [`find_first()`]: trait.ParallelIterator.html#method.find_first
+    /// [`find_any()`]: trait.ParallelIterator.html#method.find_any
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use rayon::prelude::*;
+    /// assert_eq!((0..10_000).into_par_iter()
+    ///                       .by_blocks(std::iter::successors(Some(10usize), |s| Some(s.saturating_mul(2))))
+    ///                       .find_first(|&e| e==4_999), Some(4_999))
+    /// ```
+    ///
+    /// In this example, without blocks, rayon will split the initial range into two but all work
+    /// on the right hand side (from 5,000 onwards) is **useless** since the sequential algorithm
+    /// never goes there. This means that if two threads are used there will be **no** speedup **at
+    /// all**.
+    /// `by_blocks` on the other hand will start with the leftmost range from 0 to 10, continue
+    /// with 10 to 30, the 30 to 70, 70 to 150, ...
+    /// Each subrange is treated in parallel, while all subranges are treated sequentially.
+    /// We therefore ensure a logarithmic number of blocks (and overhead) while guaranteeing
+    /// we stop at the first block containing the searched data.
+    fn by_blocks<S: IntoIterator<Item = usize>>(self, sizes: S) -> ByBlocks<Self, S> {
+        ByBlocks::new(self, sizes)
+    }
     /// Collects the results of the iterator into the specified
```

Comment:
tiny nit: please add a blank line between items like the two functions here, and also between the `type` and `fn` in your trait implementations.

Reply:
sorry i just saw this comment. in the meantime i changed the file so i'm not too sure what you are referring to.

Reply:
I'll add inline suggestions for it.

Reply:
ok this is taking shape.

however there are some functions i had which are not included here.
i'm not too sure if we should add them at least in this pull request.